### DISCUSSION TITLE
How to smooth the curve from Extra Device

### DISCUSSION THREAD

**Discussion by mansonchan888** [artisan-roaster-scope/artisan#1031] [Q&A]
I just setup  a new device for testing the vent pressure as extra device
but the curve is very unstable (like noise) (please see the light blue curve in the attached picture )
I just wonder is there any way I can set a formula (averaging a few previous data) to eliminate the noise (smooth the curve)
or some setting in Artisan software that I can try. 
I tried to set sampling to a bigger number or smooth Curve value, but the new settings will also impact  the other curves like ET & BT. The BT / ET and their delta curves are working fine for now and I don't want to impact those existing curves.

<img width="813" alt="WeChat84fc4dff4b150e649fe0ae5ba3815d82" src="https://user-images.githubusercontent.com/119805088/205503097-b09db023-7344-41c2-ab95-2b19d9be5aef.png">

**Comment by roasterdave** [comment]
Not a bad looking curve for a pressure sensor, this looks more like real data than noise.  If you still want to smooth it you will need to create a symbolic equation that computes the average value of pressure readings over some number of past samples.  Take a look at this [blog post](https://artisan-roasterscope.blogspot.com/2019/11/symbolic-formulas-basics-new-variables.html) for some examples of how to use symbolic equations.   Also check the Help button on the Config>> Device>> Extra Devices tab or view that help page in your browser [here](https://htmlpreview.github.io/?https://github.com/artisan-roaster-scope/artisan/blob/master/doc/help_dialogs/Output_html/symbolic_help.html).  A discussion of smoothing can be found in this [blog post](https://artisan-roasterscope.blogspot.com/2014/01/sampling-interval-smoothing-and-rate-of.html).

**Reply by mansonchan888** [reply]
Hi Dave
these is my finding
1) under Arduino, under y2(x) symbolic formula won't work. I only can do Y4/10 (or any simple math on Y4), not any prevous data like Y[?]. I don't know why? (bottom blue curve)
2) I can set a virtural device which using the previous values Y4[-?]  which I believe the value is the result from ArduinoTC4 56 , y2(x). please see bottom black curve (delay 7 readings)

<img width="921" alt="Screenshot 2022-12-06 at 12 52 03" src="https://user-images.githubusercontent.com/119805088/205818851-6cf69542-9298-4428-adb8-f3c237d34bf4.png">

<img width="1265" alt="Screenshot 2022-12-06 at 12 53 17" src="https://user-images.githubusercontent.com/119805088/205819034-01494544-ccb9-4302-bf09-8629ddf0661d.png">

**Comment by mansonchan888** [comment]
<img width="1298" alt="WeChat9718579559e08dd7ab94e9f61a911a38" src="https://user-images.githubusercontent.com/119805088/205524539-87d501a7-abdc-4622-8f7b-b1783e1debfe.png">

Thanks Dave, where can I put my symbolic equation for extra device?  and can you check whether my equation syntax is correct or not?  I try to put one in Device Assignment and using y2(x) = (Y4[-3]+Y4[-2]+Y4[-1]+Y4[0])/4 but it wont work..

**Reply by bigl00z3** [reply]
as you are in the Y4 field, use "x" to call the value from this field. That would be `(x[-3]+x[-2]+x[-1]+x)/4 `:

<img width="855" alt="image" src="https://user-images.githubusercontent.com/8886922/205563296-a4e0f35f-31ca-4d79-9c81-f01610d56506.png">

Nico

**Reply by mansonchan888** [reply]
Hi Nico, thanks for the reply. I tried your suggestion (x[-3]+x[-2]+x[-1]+x)/4 but it won't work. Did you try it on your side?
Please show me a picture how it work.

**Comment by MAKOMO** [comment]
If you are on y TC4 it is much better to do any smoothing computation on the Arduino itself as it loops way faster than the Artisan sampling interval. For the 4 TC4 temperature channels you can configure on-device smoothing per channel in the ET/BT tab using the "Filter" settings.

**Reply by mansonchan888** [reply]
do you mean I should setup within the device? I ask the enginerr the device was taking the input data from 2 probes (BT & ET) plus a input from pressure sensor. He told me that the data is taking per 0.25s for 3 channels. Whatever we change it will impact the data for BT and ET. if that is true, I would prefer keeping it for 0.25s sampling without impacting BT / ET.

**Reply by MAKOMO** [reply]
Your engineer has already modified the serial protocol to output your pressure sensor data in channel 5 which is by default transporting the heater duty%. He should also be able to put a running average on that data channel before sending it out to Artisan. This will be way more effective than to do smothing on the Artisan side using symbolic formulas.

**Comment by MAKOMO** [comment]
Keep the original data from the ArduinoTC4_56 without any modification (symbolic formula) and generate the smoothed data in a separate curve (device type virtual) and construct here your smoothed version of the original data. You can hide the original data and only show the smoothed one. For the virtual curve use 
`(Y4[-3]+Y4[-2]+Y4[-1]+Y4[0])/4` or whatever formula you want using data from Y4 (your Pa signal).

**Reply by MAKOMO** [reply]
.. still I do not understand why this Pa signal is not smoothed on the Arduino side (along the lines of the filters applied to channels 1-4) as you anyhow already modified the TC4 firmware...
